src/ccs.rs

```diff
@@ -0,0 +1,636 @@
+  pub(crate) ccs: CCSShape<G>,
+  // q multisets S (same as CCS)
+  // q constants c (same as CCS)
+}
+
+/// CCCS Instance is (C, x)
+/// CCCS Witness is w _mle
+
+// NOTE: We deal with `r` parameter later in `nimfs.rs` when running `execute_sequence` with `ro_consts`
+/// A type that holds a CCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCCSInstance<G: Group> {
+  // Commitment to a multilinear polynomial in s' - 1 variables
+  pub(crate) C: Commitment<G>,
+
+  // $x in F^l$
+  pub(crate) X: Vec<G::Scalar>,
+}
+
+/// A type that holds a witness for a given CCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CCCSWitness<G: Group> {
+  // Multilinear polynomial w_mle in s' - 1 variables
+  pub(crate) w_mle: Vec<G::Scalar>,
+}
+
+// NOTE: We deal with `r` parameter later in `nimfs.rs` when running `execute_sequence` with `ro_consts`
+/// A type that holds a LCCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct LCCCSInstance<G: Group> {
+  pub(crate) C: Commitment<G>,
+  pub(crate) X: Vec<G::Scalar>,
+  pub(crate) u: G::Scalar,
+  pub(crate) v: Vec<G::Scalar>,
+}
+
+/// A type that holds a witness for a given LCCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct LCCCSWitness<G: Group> {
+  pub(crate) w_mle: Vec<G::Scalar>,
+}
+
+impl<G: Group> CCS<G> {
+  // TODO: Update commitment_key variables here? This is currently based on R1CS with M length
+  /// Samples public parameters for the specified number of constraints and variables in an CCS
+  pub fn commitment_key(S: &CCSShape<G>) -> CommitmentKey<G> {
+    let num_cons = S.num_cons;
+    let num_vars = S.num_vars;
+    let total_nz = S.M.iter().fold(0, |acc, m| acc + m.0.len());
+
+    G::CE::setup(b"ck", max(max(num_cons, num_vars), total_nz))
+  }
+}
+
+impl<G: Group> CCSShape<G> {
+  /// Create an object of type `CCSSShape` from the explicitly specified CCS matrices
+  pub fn new(
+    num_cons: usize,
+    num_vars: usize,
+    num_io: usize,
+    M: &[Vec<(usize, usize, G::Scalar)>],
+    t: usize,
+    q: usize,
+    d: usize,
+    S: Vec<Vec<usize>>,
+    c: Vec<G::Scalar>,
+  ) -> Result<CCSShape<G>, NovaError> {
+    let is_valid = |num_cons: usize,
+                    num_vars: usize,
+                    num_io: usize,
+                    matrix: &[(usize, usize, G::Scalar)]|
+     -> Result<(), NovaError> {
+      let res = (0..matrix.len())
+        .map(|i| {
+          let (row, col, _val) = matrix[i];
+          if row >= num_cons || col > num_io + num_vars {
+            Err(NovaError::InvalidIndex)
+          } else {
+            Ok(())
+          }
+        })
+        .collect::<Result<Vec<()>, NovaError>>();
+
+      if res.is_err() {
+        Err(NovaError::InvalidIndex)
+      } else {
+        Ok(())
+      }
+    };
```

Comment:
Can't we have this as a sepparated fn for `CCSShape`? It would make more sense. And we can make it `pub(crate)` and play more with debug features or similar.